app/services/report_importer.rb

```diff
@@ -91,21 +91,60 @@ def logs
   end
 
   def import_log_messages
+    id_messages = []
+    db_messages = []
+    id_sources = []
+    db_sources = []
+    db_logs = []
     logs.each do |log|
-      # Parse the API format
+      msg = log['log']['messages']['message']
+      src = log['log']['sources']['source']
+
+      msg_id = Message.where(value: msg).pluck(:id).first
```

Comment:
This does a ton of database queries which is typically very slow, especially on remote databases. It's probably faster to do a single DB query for all messages and see which already exist. Same for Source below.

Reply:
Yes I was thinking the same, but you can have thousands of facts (we've seen 10,000 facts for a host with 1400 HDDs) and this will definitely break some limit of SQL query. I'd probably need to break this into some kind of batches, this will make the code even more complex. Do we really want to do that?

On the other hand, we can set some maximum mount of facts. Not sure how a host with 10,000 facts is actually useful. Postgres also has quite reasonable length limits and it appears to be configurable. Let me think about this and try some more testing if it helps to improve performance even further.

Reply:
We shouldn't mix up facts and reports. This is just reports and reports have no facts. Messages shouldn't be dropped.

A report can have 100s or 1000s of messages. In that case, we will do 1000s of SQL queries and that's known to be slow.

However, we also know that a huge `WHERE IN ( .... )` isn't great. You can consider temporary tables and join.

Reply:
Sorry disregard my comment about facts, I find hard to switch from facts to reports. I often work on these areas :-)

Anyway, a report can have many lines therefore we would generate a huge SQL statement. I am not sure about temporary tables if it is worth the effort. I need to take a look.

Reply:
I was looking into possibility of a single query and I cannot do it. The idea behind this optimization is to find out if an existing message/source is already present. I cannot do this in a single call because SQL servers randomize output when you use "WHERE ... IN" statement. There are some really complex solutions (using special sorted arrays) for postgres, but I don't think this is worth the effort.

Another problem is that the SQL query contains whole message/source line as the hash is now calculated in the SQL driver/server. Therefore this can make the single query very long, we are not transferring just IDs but long lines.

The patch still shows good improvement, we used to do ton of queries and then ton of insert statements. Now it leverages `insert_all` Rails method which performs inserting in batches. This is very important specifically for the join table (logs).